ducktape/cluster/node_container.py

```diff
@@ -103,25 +122,58 @@ def remove_nodes(self, nodes):
         for node in nodes:
             self.remove_node(node)
 
-    def remove_spec(self, cluster_spec):
+    def remove_spec(self, cluster_spec) -> RemoveSpecResult:
         """
         Remove nodes matching a ClusterSpec from this NodeContainer.
 
         :param cluster_spec:                    The cluster spec.  This will not be modified.
-        :returns:                               A list of the nodes that were removed.
-        :throws InsufficientResourcesError:     If there are not enough nodes in the NodeContainer.
-                                                Nothing will be removed unless enough are available.
+        :returns:                               An instance of `AllocResult`, which contains good nodes that passed
+                                                health checks, bad nodes that didn't, overall success/failure and
+                                                a detailed message (if any).
+        :raises:                                InsufficientResourcesError when there aren't enough total nodes
+                                                InsufficientHealthyNodesError when there aren't enough healthy nodes
         """
-        msg = self.attempt_remove_spec(cluster_spec)
-        if len(msg) > 0:
-            raise InsufficientResourcesError("Not enough nodes available to allocate. " + msg)
-        removed = []
+
+        err = self.attempt_remove_spec(cluster_spec)
+        if err:
+            # there weren't enough nodes to even attempt allocations, raise the exception
+            raise InsufficientResourcesError(err)
```

Comment:
This is exactly how it was before. I'm agreeing on this being go-like, but I don't want to refactor other functions just because of that, at least not in this PR - it's already too big.